pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,771 @@
+			if err != nil {
+				return fmt.Errorf("couldn't write packets: %w", err)
+			}
+			klog.V(5).InfoS("Capture packets", "name", captureState.name, "count",
+				captureState.capturedPacketsNum, "len", ci.Length)
+
+			c.lock.Lock()
+			reachTarget := captureState.isCaptureSucceed()
+			c.lock.Unlock()
+			// use rate limiter to reduce the times we need to update status.
+			if reachTarget || captureState.updateRateLimiter.Allow() {
+				pc, err := c.packetCaptureLister.Get(captureState.name)
+				if err != nil {
+					return fmt.Errorf("get PacketCapture failed: %w", err)
+				}
+				// if reach the target. flush the file and upload it.
+				if reachTarget {
+					path := env.GetPodName() + ":" + nameToPath(pc.Name)
+					statusPath := path
+					if err = captureState.pcapngWriter.Flush(); err != nil {
+						return err
+					}
+					if pc.Spec.FileServer != nil {
+						err = c.uploadPackets(pc, captureState.pcapngFile)
+						klog.V(4).InfoS("Upload captured packets", "name", pc.Name, "path", path)
+						statusPath = fmt.Sprintf("%s/%s.pcapng", pc.Spec.FileServer.URL, pc.Name)
+					}
+					c.lock.Lock()
+					captureState.filePath = statusPath
+					c.lock.Unlock()
+					if err != nil {
+						return err
+					}
+					if err := captureState.pcapngFile.Close(); err != nil {
+						klog.ErrorS(err, "Close pcapng file error", "name", pc.Name, "path", path)
+					}
+				}
+				// report capture status.
+				c.addPacketCapture(pc)
+			}
+		case <-ctx.Done():
+			pc, err := c.packetCaptureLister.Get(captureState.name)
+			if err != nil {
+				return fmt.Errorf("get PacketCapture failed: %w", err)
+			}
+			klog.InfoS("PacketCapture timeout", "name", pc.Name)
+			return errors.New(captureTimeoutReason)
+		}
+	}
+}
+
+func (c *Controller) getPodIP(podRef *crdv1alpha1.PodReference) (net.IP, error) {
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(podRef.Name, podRef.Namespace)
+	var podIP net.IP
+	if len(podInterfaces) > 0 {
+		podIP = podInterfaces[0].GetIPv4Addr()
+	} else {
+		pod, err := c.kubeClient.CoreV1().Pods(podRef.Namespace).Get(context.TODO(), podRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get Pod %s/%s: %w", podRef.Namespace, podRef.Name, err)
+		}
+		podIPs := make([]net.IP, len(pod.Status.PodIPs))
+		for i, ip := range pod.Status.PodIPs {
+			podIPs[i] = net.ParseIP(ip.IP)
+		}
+		podIP = util.GetIPv4Addr(podIPs)
+	}
+	if podIP == nil {
+		return nil, fmt.Errorf("cannot find IP with IPv4 address family for Pod %s/%s", podRef.Namespace, podRef.Name)
+	}
+	return podIP, nil
+}
+
+func (c *Controller) parseIPs(pc *crdv1alpha1.PacketCapture) (srcIP, dstIP net.IP, err error) {
+	if pc.Spec.Source.Pod != nil {
+		srcIP, err = c.getPodIP(pc.Spec.Source.Pod)
+	} else if pc.Spec.Source.IP != nil {
+		srcIP = net.ParseIP(*pc.Spec.Source.IP)
+		if srcIP == nil {
+			err = fmt.Errorf("invalid source IP address: %s", *pc.Spec.Source.IP)
```

Comment:
The two err would be overriden by the following err.

Reply:
added early  return.